pkg/media/opuswriter/opus-writer.go

```diff
@@ -0,0 +1,190 @@
+
+       Page 0         Pages 1 ... n        Pages (n+1) ...
+    +------------+ +---+ +---+ ... +---+ +-----------+ +---------+ +--
+    |            | |   | |   |     |   | |           | |         | |
+    |+----------+| |+-----------------+| |+-------------------+ +-----
+    |||ID Header|| ||  Comment Header || ||Audio Data Packet 1| | ...
+    |+----------+| |+-----------------+| |+-------------------+ +-----
+    |            | |   | |   |     |   | |           | |         | |
+    +------------+ +---+ +---+ ... +---+ +-----------+ +---------+ +--
+    ^      ^                           ^
+    |      |                           |
+    |      |                           Mandatory Page Break
+    |      |
+    |      ID header is contained on a single page
+    |
+    'Beginning Of Stream'
+
+   Figure 1: Example Packet Organization for a Logical Ogg Opus Stream
+*/
+
+// New builds a new OGG Opus writer
+func New(fileName string, sampleRate uint32, channelCount uint16) (*Writer, error) {
+	f, err := os.Create(fileName)
+	if err != nil {
+		return nil, err
+	}
+	// Since the seed and the creation date of the
+	// file will be really close, is this an issue ?
+	r := rand.NewSource(time.Now().UnixNano())
+	r1 := rand.New(r)
+	serial := uint32(math.Ceil(r1.Float64() * math.Pow(2, 32)))
+	writer := &Writer{
+		fd:            f,
+		sampleRate:    sampleRate,
+		channelCount:  channelCount,
+		serial:        serial,
+		checksumTable: make([]int, 256),
+	}
+	writer.initChecksum()
+
+	// ID Header
+	oggIDHeader := make([]byte, 19)
+
+	copy(oggIDHeader[0:], []byte("OpusHead"))                   // Magic Signature 'OpusHead'
+	oggIDHeader[8] = 1                                          // Version
+	oggIDHeader[9] = uint8(channelCount)                        // Channel count
+	binary.LittleEndian.PutUint16(oggIDHeader[10:], 0)          // pre-skip, don't need to skip any value
+	binary.LittleEndian.PutUint32(oggIDHeader[12:], sampleRate) // original sample rate, any valid sample e.g 48000
+	binary.LittleEndian.PutUint16(oggIDHeader[16:], 0)          // output gain
+	oggIDHeader[18] = 0                                         // channel map 0 = one stream: mono or stereo
+
+	// RFC specifies that the ID Header page should have a granule position of 0 and a Header Type set to 2 (StartOfStream)
+	data := writer.createPage(oggIDHeader, 2, 0)
+	if _, err := f.Write(data); err != nil {
+		return nil, err
+	}
+
+	// Comment Header
+	oggCommentHeader := make([]byte, 21)
+	copy(oggCommentHeader[0:], []byte("OpusTags"))          // Magic Signature 'OpusTags'
+	binary.LittleEndian.PutUint32(oggCommentHeader[8:], 5)  // Vendor Length
+	copy(oggCommentHeader[12:], []byte("pions"))            // Vendor name 'pions'
+	binary.LittleEndian.PutUint32(oggCommentHeader[17:], 0) // User Comment List Length
+
+	// RFC specifies that the page where the CommentHeader completes should have a granule position of 0
```

Comment:
Is it easy to link to the RFC here? Just nice for the reader, if it is a PDF or anything don't worry about it!

Reply:
I used an online RFC, I can add the link to it :)